src/cdk/a11y/focus-trap/focus-trap-manager.ts

```diff
@@ -0,0 +1,61 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Injectable} from '@angular/core';
+
+/**
+ * A FocusTrap managed by FocusTrapManager.
+ * Implemented by ConfigurableFocusTrap to avoid circular dependency.
+ */
+export interface ManagedFocusTrap {
+    enable(): void;
+    disable(): void;
+    focusInitialElementWhenReady(): void;
+}
+
+/** Injectable that ensures only the most recently enabled FocusTrap is active. */
+@Injectable({providedIn: 'root'})
+export class FocusTrapManager {
+  // A stack of the FocusTraps on the page. Only the FocusTrap at the
+  // top of the stack is active.
+  private _focusTrapStack: ManagedFocusTrap[] = [];
+
+  /**
+   * Disables the FocusTrap at the top of the stack, and then pushes
+   * the new FocusTrap onto the stack.
+   */
+  register(focusTrap: ManagedFocusTrap): void {
+    // Dedupe focusTraps that register multiple times.
+    this._focusTrapStack = this._focusTrapStack.filter(function(ft) {
+      return ft !== focusTrap;
+    });
+
+    if (this._focusTrapStack.length) {
+      this._focusTrapStack[this._focusTrapStack.length - 1].disable();
+    }
+
+    this._focusTrapStack.push(focusTrap);
+    focusTrap.enable();
+  }
+
+  /**
+   * Removes the FocusTrap from the stack, and activates the
+   * FocusTrap that is the new top of the stack.
+   */
+  unregister(focusTrap: ManagedFocusTrap): void {
+    focusTrap.disable();
+    const i = this._focusTrapStack.indexOf(focusTrap);
+    if (i !== -1) {
+      this._focusTrapStack.splice(i, 1);
```

Comment:
Since you're repeating `this._focusTrapStack` a bunch of times, it might be worth pulling it out into a variable.